File: internal/translation/google.go
package translation

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
)

type GoogleTranslator struct {
	client *http.Client
}

func NewGoogle() *GoogleTranslator {
	return &GoogleTranslator{client: &http.Client{}}
}

func (g *GoogleTranslator) Translate(ctx context.Context, req Request) (string, error) {
	sl, tl := "en", "ja"
	if req.Direction == Outbound {
		sl, tl = "ja", "en"
	}

	u := "https://translate.googleapis.com/translate_a/single?" + url.Values{
		"client": {"gtx"},
		"sl":     {sl},
		"tl":     {tl},
		"dt":     {"t"},
		"q":      {req.Message},
	}.Encode()

	httpReq, err := http.NewRequestWithContext(ctx, http.MethodGet, u, nil)
	if err != nil {
		return "", err
	}
	httpReq.Header.Set("User-Agent", "Mozilla/5.0")

	resp, err := g.client.Do(httpReq)
	if err != nil {
		return "", fmt.Errorf("google translate request: %w", err)
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return "", fmt.Errorf("google translate read: %w", err)
	}

	if resp.StatusCode != http.StatusOK {
		return "", fmt.Errorf("google translate status %d: %s", resp.StatusCode, body)
	}

	return parseGoogleResponse(body)
}

func parseGoogleResponse(body []byte) (string, error) {
	var data []any
	if err := json.Unmarshal(body, &data); err != nil {
		return "", fmt.Errorf("google translate parse: %w", err)
	}

	sentences, ok := data[0].([]any)
	if !ok {
		return "", fmt.Errorf("google translate: unexpected response format")
	}

	var result string
	for _, s := range sentences {
		parts, ok := s.([]any)
		if !ok || len(parts) == 0 {
			continue
		}
		if text, ok := parts[0].(string); ok {
			result += text
		}
	}
	return result, nil
}

File: src-tauri/src/translation/google.rs
use super::TranslationRequest;

pub async fn translate(
    client: &reqwest::Client,
    req: &TranslationRequest,
) -> Result<String, String> {
    let (sl, tl) = if req.direction == super::Direction::Outbound {
        (req.target_lang.as_str(), "en")
    } else {
        ("auto", req.target_lang.as_str())
    };

    let url = reqwest::Url::parse_with_params(
        "https://translate.googleapis.com/translate_a/single",
        &[
            ("client", "gtx"),
            ("sl", sl),
            ("tl", tl),
            ("dt", "t"),
            ("q", req.message.as_str()),
        ],
    )
    .map_err(|e| format!("google translate url: {}", e))?;

    let resp = client
        .get(url)
        .header("User-Agent", "Mozilla/5.0")
        .send()
        .await
        .map_err(|e| format!("google translate request: {}", e))?;

    let status = resp.status();
    let body = resp
        .text()
        .await
        .map_err(|e| format!("google translate read: {}", e))?;

    if !status.is_success() {
        return Err(format!("google translate status {}: {}", status, body));
    }

    parse_response(&body)
}

fn parse_response(body: &str) -> Result<String, String> {
    let data: serde_json::Value =
        serde_json::from_str(body).map_err(|e| format!("google translate parse: {}", e))?;

    let sentences = data
        .as_array()
        .and_then(|arr| arr.first())
        .and_then(|v| v.as_array())
        .ok_or_else(|| "google translate: unexpected response format".to_string())?;

    let mut result = String::new();
    for sentence in sentences {
        if let Some(text) = sentence
            .as_array()
            .and_then(|parts| parts.first())
            .and_then(|v| v.as_str())
        {
            result.push_str(text);
        }
    }

    Ok(result)
}
